src/backend/optimizer/plan/transform.c

```diff
@@ -520,3 +521,134 @@ replace_sirvf_rte(Query *query, RangeTblEntry *rte)
 
 	return rte;
 }
+
+/*
+ * Does query has SRFs, or WITH ORDINALITY?
+ */
+bool query_has_srf(Query *query)
+{
+	if (query->hasTargetSRFs)
+	{
+		return true;
+	}
+
+	/* Double check for subquery. */
+	if (expression_returns_set( (Node *) query->targetList))
+	{
+		return true;
+	}
+
+	ListCell *lcrte = NULL;
+	foreach (lcrte, query->rtable)
+	{
+		RangeTblEntry *rte = (RangeTblEntry *) lfirst(lcrte);
+
+		switch(rte->rtekind)
+		{
+			case RTE_FUNCTION:
+			{
+				ListCell *lcrtfunc;
+
+				/* WITH ORDINALITY */
+				if (rte->funcordinality)
+					return true;
+
+				foreach(lcrtfunc, rte->functions)
+				{
+					RangeTblFunction *rtfunc = (RangeTblFunction *) lfirst(lcrtfunc);
+
+					if (!IsA(rtfunc->funcexpr, FuncExpr))
+						return true;
+
+					if (((FuncExpr *) rtfunc->funcexpr)->funcretset)
+					{
+						/* SRF in FROM clause */
+						return true;
+					}
+				}
+				break;
+			}
+			case RTE_SUBQUERY:
+			{
+				Query *sq = (Query *) rte->subquery;
+
+				if (query_has_srf(sq))
+				{
+					return true;
+				}
+				break;
+			}
+			default:
+			{
+				break;
+			}
+		}
+	}
+
+	return false;
+}
+
+/*
+ * DISTINCT/DISTINCT ON/ORDER BY optimization.
+ * Remove DISTINCT clause if possibile, ex:
+ * select DISTINCT count(a) from t; to
+ * select count(a) from t;
+ * There is one row returned at most, DISTINCT and/or ON is pointless then.
+ * The same with ORDER BY clause;
+ */
+Query *remove_distinct_sort_clause(Query *parse)
+{
+	if (parse->hasAggs &&
+		parse->groupClause == NIL &&
+		!contain_mutable_functions((Node *) parse) &&
+		!query_has_srf(parse))
+	{
+		List	   *useless_tlist = NIL;
+		List	   *tles;
+		List	   *sortops;
+		List	   *eqops;
+		ListCell   *lc;
+
+		if (parse->distinctClause != NIL)
```

Comment:
Hmm... If I understand correctly, `query_has_srf` doesn't need to care about subquery or child nodes, it only need to check if the targetlist contains any srf. (the targetlis which contains distinct expression).
But for `remove_distinct_sort_clause`, seems it only process the topmost targetlist, subqueries are not processed
such as :
```
select * from t, (select distinct(count(*) from t);
```

Reply:
> select * from t, (select distinct(count(*) from t);

 Does this SQL return the wrong results or unexpected plan, could you show the results?

Reply:
> select * from t, (select distinct(count(*) from t);

That seems an invalid SQL:
![image](https://github.com/user-attachments/assets/0ca7bff9-c5ae-4852-b501-24f97922874b)

Reply:
```
select * from t, (select distinct(count(*) from t);
```
for the above query, the distinct can be removed, as  the 'remove_distinct_sort_clause' can be applied on each subquery, don't need call it recursively, I misunderstood that.

but for the below query 
```
explain (verbose) select distinct(count(t.a)) from t,  (select distinct(count(*)), generate_series(0,1))as xx;
                                                    QUERY PLAN
------------------------------------------------------------------------------------------------------------------
 Unique  (cost=10000001318.17..10000001318.18 rows=1 width=8)
   Output: (count(t.a))
   Group Key: (count(t.a))
   ->  Sort  (cost=10000001318.17..10000001318.17 rows=1 width=8)
         Output: (count(t.a))
         Sort Key: (count(t.a))
         ->  Finalize Aggregate  (cost=10000001318.15..10000001318.16 rows=1 width=8)
               Output: count(t.a)
               ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=10000001318.09..10000001318.14 rows=3 width=8)
                     Output: (PARTIAL count(t.a))
                     ->  Partial Aggregate  (cost=10000001318.09..10000001318.10 rows=1 width=8)
                           Output: PARTIAL count(t.a)
                           ->  Nested Loop  (cost=10000000000.05..10000001157.59 rows=64200 width=4)
                                 Output: t.a
                                 ->  Seq Scan on public.t  (cost=0.00..355.00 rows=32100 width=4)
                                       Output: t.a
                                 ->  Materialize  (cost=0.05..0.10 rows=2 width=0)
                                       ->  Subquery Scan on xx  (cost=0.05..0.09 rows=2 width=0)
                                             ->  Unique  (cost=0.05..0.07 rows=2 width=12)
                                                   Output: (count(*)), (generate_series(0, 1))
                                                   Group Key: (count(*)), (generate_series(0, 1))
                                                   ->  Sort  (cost=0.05..0.06 rows=2 width=12)
                                                         Output: (count(*)), (generate_series(0, 1))
                                                         Sort Key: (count(*)), (generate_series(0, 1))
                                                         ->  ProjectSet  (cost=0.01..0.04 rows=2 width=12)
                                                               Output: (count(*)), generate_series(0, 1)
                                                               ->  Aggregate  (cost=0.01..0.02 rows=1 width=8)
                                                                     Output: count(*)
                                                                     ->  Result  (cost=0.00..0.01 rows=1 width=0)
 Settings: optimizer = 'off'
 Optimizer: Postgres query optimizer
(31 rows)
```
the `distinct(count(t.a))` is not optimized.

Reply:
> explain (verbose) select distinct(count(t.a)) from t,  (select distinct(count(*)), generate_series(0,1))as xx;
> the distinct(count(t.a)) is not optimized.

Good catch. This was blocked by SRF checks as subquery has SRF.
`distinct(count(t.a))` seems ok to be optimized if we only care the TargetList and ignore other parts of Query.

For`distinct(count(*))` it's not easy to remove because pg planner build everything from end to top, however ORCA could do that using Cascade Model.

Reply:
> distinct(count(t.a)) seems ok to be optimized if we only care the TargetList and ignore other parts of Query.

Fixed as that, thanks for catch! @fanfuxiaoran